add flow-informed tuning guidance (#16265)
elastic/logstash: docs/static/performance-checklist.asciidoc

Depending on the input plugin being used, back-pressure can either propagate upstream or result in the risk of data-loss.

A pipeline's `queue_backpressure` flow metric reflects how much time the inputs are spending attempting to push events into the queue.
The metric isn't precisely comparable across pipelines, but instead allows you to compare a single pipeline's current behaviour to _itself_ over time.
When this metric is growing, we generally need to look _downstream_ at the pipeline's filters and outputs to see if they are using resources effectively, have sufficient resources allocated, or are experiencing back-pressure of their own.

NOTE: A persisted queue offers durability guarantees and can absorb back-pressure for longer than the default in-memory queue, but once it is full it too propagates back-pressure.
      The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.

> [[tuning-logstash]]

Author: It is not clear to me why this section-break results in a separate page. I had hoped that the bits above (the why) would be more closely-tied to the bits below (the how).

Is it just the heading level, or is there something else?

Reviewer: Is the presentation you want similar to this?  If so, the `[float]` tag is the magic you're looking for. 

<img width="1105" alt="Screen Shot 2024-06-27 at 4 58 59 PM" src="https://github.com/elastic/logstash/assets/35154725/a8c04b03-2763-46c1-a545-7e25f9918044">

>       The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.
> 
> [[tuning-logstash]]

Reviewer: ```suggestion
      The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.

[float]
[[tuning-logstash]]
```

Author: After a separate conversation in slack, I think I actually need to move the content to be _inside_ of the next heading, to set context before that page gets into the specifics.

Thank you! Knowing that `[float]` means keep-with-previous helps a lot.

When this is the case, we can see _which_ plugins in the pipeline are consuming the available worker capacity by looking at the plugin-level `worker_utilization` and `worker_millis_per_event` flow metrics.
Using this information, we can gain intuition about how to tune the pipeline's settings to add resources, or how to find and eliminate wasteful computation.

In general, plugins fit into one of two categories:

- *CPU-bound*: plugins that perform computation on the contents of events _without_ the use of the network or disk IO tend to benefit from incrementally increasing `pipeline.workers` as long as the process has available CPU; once CPU is exhausted additional concurrency can result in _lower_ throughput as the pipeline workers contend for resources.
- *IO-bound*: plugins that use the network to either enrich events or transmit events tend to benefit from incrementally increasing `pipeline.workers` and/or tuning the pipeline batching as long as those external services are not exerting back-pressure (even if Logstash is using nearly all of its available CPU).

> When a pipeline's `worker_utilization` is _below_ 100, its pipeline workers are spending their time waiting for input.

Author: ```suggestion
The further a pipeline's `worker_utilization` is from 100, the more time its workers are spending waiting for events to arrive in the queue.
```

[float]
[[tuning-logstash-queue-backpressure]]
==== Queue back-pressure

When a pipeline receives events faster than it can process them, the inputs eventually experience back-pressure that prevents them from receiving additional events.
Depending on the input plugin being used, back-pressure can either propagate upstream or result in the risk of data-loss.

A pipeline's `queue_backpressure` flow metric reflects how much time the inputs are spending attempting to push events into the queue.
The metric isn't precisely comparable across pipelines, but instead allows you to compare a single pipeline's current behaviour to _itself_ over time.
When this metric is growing, we generally need to look _downstream_ at the pipeline's filters and outputs to see if they are using resources effectively, have sufficient resources allocated, or are experiencing back-pressure of their own.

NOTE: A persisted queue offers durability guarantees and can absorb back-pressure for longer than the default in-memory queue, but once it is full it too propagates back-pressure.

> The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.

Author: ```suggestion
The `queue_persisted_growth_events` flow metric is useful measure of how much back-pressure is being actively absorbed by the persisted queue, and should trend toward zero (or less) over the pipeline's lifetime.
Negative numbers indicate that the queue is _shrinking_, and that the workers are catching up on lag that had previously developed.
```

[float]
[[tuning-logstash-settings]]
==== Tuning-related settings

The Logstash defaults are chosen to provide fast, safe performance for most users.
However if you notice performance issues, you may need to modify some of the defaults.
Logstash provides the following configurable options for tuning pipeline performance: `pipeline.workers`, `pipeline.batch.size`, and `pipeline.batch.delay`.

For more information about setting these options, see <<logstash-settings-file>>.

Make sure you've read the <<performance-troubleshooting>> before modifying these options.

* The `pipeline.workers` setting determines how many threads to run for filter and output processing.
  If you find that events are backing up, or that the CPU is not saturated, consider increasing the value of this parameter to make better use of available processing power.
  Good results can even be found increasing this number past the number of available processors as these threads may spend significant time in an I/O wait state when writing to external systems.

>   Legal values for this parameter are positive integers.

Author: ```suggestion
```

I agree -- it was here because of the net-no-change refactor to support the one-sentence-per-line convention, but can be ✂️ here.

* The `pipeline.batch.size` setting defines the maximum number of events an individual worker thread collects before attempting to execute filters and outputs.
  Larger batch sizes are generally more efficient, but increase memory overhead.

>   Some hardware configurations require you to increase JVM heap space in the `jvm.options` config file to avoid performance degradation (See <<config-setting-files>> for more info).
>   Values in excess of the optimum range cause performance degradation due to frequent garbage collection or JVM crashes related to out-of-memory exceptions.

Author: They really do. I'll snip.

```suggestion
```

* The number of workers may be set higher than the number of CPU cores since outputs often spend idle time in I/O wait conditions.

* Threads in Java have names and you can use the `jstack`, `top`, and the VisualVM graphical tools to figure out which resources a given thread uses.

* On Linux platforms, Logstash labels all the threads it can with something descriptive.
  For example, inputs show up as `[base]<inputname`, and pipeline workers show up as `[base]>workerN`, where N is an integer.
  Where possible, other threads are also labeled to help you identify their purpose.

[float]
[[profiling-the-heap]]
==== Profiling the heap

When tuning Logstash you may have to adjust the heap size.
You can use the https://visualvm.github.io/[VisualVM] tool to profile the heap.
The *Monitor* pane in particular is useful for checking whether your heap allocation is sufficient for the current workload.
The screenshots below show sample *Monitor* panes.
The first pane examines a Logstash instance configured with too many inflight events.
The second pane examines a Logstash instance configured with an appropriate amount of inflight events.

> Note that the specific batch sizes used here are most likely not applicable to your specific workload, as the memory demands of Logstash vary in large part based on the type of messages you are sending.

Reviewer 2: not worth changing now, but typically I prefer to put style changes in a separate PR to facilitate reviewing, especially in documentation where github isn't great at showing diffs.
